server/src/main/java/org/opensearch/search/SearchService.java

```diff
@@ -1525,19 +1529,40 @@ private CanMatchResponse canMatch(ShardSearchRequest request, boolean checkRefre
                 final boolean aliasFilterCanMatch = request.getAliasFilter().getQueryBuilder() instanceof MatchNoneQueryBuilder == false;
                 FieldSortBuilder sortBuilder = FieldSortBuilder.getPrimaryFieldSortOrNull(request.source());
                 MinAndMax<?> minMax = sortBuilder != null ? FieldSortBuilder.getMinMaxOrNull(context, sortBuilder) : null;
-                final boolean canMatch;
+                boolean canMatch;
                 if (canRewriteToMatchNone(request.source())) {
                     QueryBuilder queryBuilder = request.source().query();
                     canMatch = aliasFilterCanMatch && queryBuilder instanceof MatchNoneQueryBuilder == false;
+                    canMatch = aliasFilterCanMatch && queryBuilder instanceof MatchNoneQueryBuilder == false;
                 } else {
                     // null query means match_all
                     canMatch = aliasFilterCanMatch;
                 }
+                canMatch = canMatch & canMatchSearchAfter(searchContext, minMax);
```

Comment:
I believe you meant a logical AND (`&&`). Bitwise AND (`&`) on a boolean doesn't short-circuit, so it will always end up executing `canMatchSearchAfter` even when `canMatch` is already false.

Reply:
Oh nice, catch !